### DISCUSSION TITLE
Monorepo data structure rules

### DISCUSSION THREAD

**Discussion by tukusejssirs** [romcal/romcal#528] [Ideas]
## @tukusejssirs’ proposals

Here are my suggestions/notes regarding (not only) data structure rules:

- Non-l10n data should be locale-agnostic:
   - ATM, [`martyrology.ts`](https://github.com/romcal/romcal/blob/dc48f77f5b9efebcb6c79c07d2b660af7f2f049e/lib/catalog/martyrology.ts#L8) contains localised `name`.
- The data sources should be written in TypeScript, compilable to both JSON and YAML, incl a JSON schema (the JSON schema could be published to [json-schema.org](https://json-schema.org/)).
- ATM, [my suggested TS data types](https://github.com/romcal/romcal/discussions/469#discussioncomment-9099898) are mostly types with some interfaces, however, some of them should be classes (e.g. calendar, celebration, source) with some methods (e.g. `Calendar`/`Source` should have a `new()`/`add()` public method; there should be some private methods to generate IDs and other stuff):
   - We could use `new Celebration(config)` instead which would be added to a `Calendar` instance (e.g. `const cal = new Calendar(config); cal.addCelebration(new Celebration(config))`; we might also support adding multiple celebrations at once, e.g. `cal.addCelebrations([new Celebration(config)])`).
- I suggest putting unit tests near the tested source files with `*.spec.ts` extension suffix and keep them at the source file they below to (e.g. `some-file.suffix.ts` → `some-file.suffix.spec.ts`.
- I suggest putting a single exported type/object/class/whatever in a file.
   - ATM, we export multiple stuff from a single file, e.g. [here](https://github.com/romcal/romcal/blob/dc48f77f5b9efebcb6c79c07d2b660af7f2f049e/lib/types/locale.ts).
   - I also suggest adding an extension suffix, e.g. `*.type.ts`, `*.interface.ts`, `*.enum.ts`, `*.constant.ts` (for export const VAR_NAME = ..., `*.dto.ts`, `*.function.ts`, possibly `*.class.ts` (although a class should have an extension suffix derived from usage, e.g. `*.dto.ts`, `*.service.ts`).
- No data should be added without referencing its source.
   - By data I mean:
      - celebration metadata (ATM it is mostly located in `martyrology.ts`):
         - the celebration metadata must be fully locale-agnostic;
      - localisation;
      - calendars with celebrations.
- No code should be added without creating a unit test for it.
- We should consider adding e2e tests.
- No user-facing code/feature/change should be added without documenting it.
- Documentation should be a separate app (in `apps/docs`), although we could keep it internal/non-publishable).

/cc @TobiTenno, @emagnier

## Resolutions and accepted stuff

I'm trying to summarize our discussions on this topic: 

1. **Build & Export Format**:
   - Agreement to distribute packages as JS files for simplicity and native typing (CJS, ESM, IIFE formats).
   - Potentially consider adding a 4th format in JSON (+ JSON Schema).

2. **Distribution**:
   - Currently distributed as separate NPM packages (keeping this method for now).

3. **Technical Considerations**:
    - Exploring the use of a tree-shaker tool to eliminate unnecessary content in localized calendar packages could significantly simplify the build process, as opposed to constructing them as currently done.
   - Add some tests to ensure the build and packaging processes work correctly.
   - Data validation itself should also be performed, but separately from the build + packaging process (separation of concerns).

**Comment by TobiTenno** [comment]
Initial pass, the only main part that i disagree with is putting the spec files in line with the files, as this can make several processing rules more complicated than they need to be.

I think the suffixes can go a bit too far, like I am worried that these suffixes would be better as just different folders or packages entirely. If the suffixes would be different aspects within a single calendar export, for example, then i could understand doing that to avoid having different folders all with one file, but for things like enums and constants that we already group in their own folders already, i don't feel that's necessary here.

How i'm reading your suggestion of what would boil down to multiple suffixes (`some-file.suffix.spec.ts`) furthers the point a bit, because we end up making the file name long enough that the imports become a mess... and the imports are largely auto-managed, but i still want to avoid this happening for consumers, depending on how exactly we expose the calendars as well.

**Reply by tukusejssirs** [reply]
> Initial pass, the only main part that i disagree with is putting the spec files in line with the files, as this can make several processing rules more complicated than they need to be.

Well, which the spec files location is not a hard requirement from my side, however, I do think it is better than looking for a test file in one location and for the code it tests in another.

Anyway, this is (at least implicitly) suggested/preferred by Nx and Nest. Here some (less than ideal) examples:

- https://github.com/nrwl/nx-examples/tree/master/apps/cart/src/app;
- https://github.com/nrwl/nx-examples/tree/master/apps/products/src/app.


> I think the suffixes can go a bit too far, like I am worried that these suffixes would be better as just different folders or packages entirely. If the suffixes would be different aspects within a single calendar export, for example, then i could understand doing that to avoid having different folders all with one file, but for things like enums and constants that we already group in their own folders already, i don't feel that's necessary here.

Nest prefers using suffices like `*.controller.ts`, `*.service.ts`, `*.dto.ts`, `*.resolver.ts`, `*.module.ts` and others. Files with these (and other) suffices are generated using Nest CLI.

Again, I don’t consider this a hard requirement, however, I think it is a good thing to follow. I personally don’t care much about filename lengths.

However, I usually move the files into separate folder by suffix, as otherwise I would have zillion files in a Nest module folder, thus it makes sense to me to move some folders into subfolders. Again, it is not a hard required.

> How i'm reading your suggestion of what would boil down to multiple suffixes (`some-file.suffix.spec.ts`) furthers the point a bit, because we end up making the file name long enough that the imports become a mess... and the imports are largely auto-managed, but i still want to avoid this happening for consumers, depending on how exactly we expose the calendars as well.

Re multiple suffices: for code files, there would a single suffix, for unit test files there would a second suffix placed between code file suffix (e.g. `enum`) and file extension (`spec`). No other suffix level would be used.

**Comment by TobiTenno** [comment]
re: Names in martryology:
would anyone be opposed to either (1) moving these names to the locale files and expanding locales to be objects w/ `name`, `titledName`, and any other combinations we need to create over time; or (2) using the localized value from the locales files entirely and remove `martyrology#name` entirely?

**Reply by emagnier** [reply]
That could indeed be a good idea (1+2) 👍 

We should establish a fallback system (similar to how it's already done with English) so that non-directly translatable metadata (like boolean data or dates, for example, such as the current `dateOfDeathIsApproximative: boolean` property) can automatically propagate into all bundles (packages containing calendar definitions and associated localization) generated during the library build.

However, it's important to note that the `titles` properties are currently used by Romcal to determine the liturgical color of the day (if the person is a martyr = red, otherwise white, or violet depending on the liturgical period). We need to figure out how to address this issue and find the right balance: either redefine titles or colors at the calendar definitions level (potentially duplicating the same metadata for the same entity); or preferably, look for this information in the localization file without requiring a localization library to fetch this specific information (I want to ensure that a localization library is not necessary for the core functioning of Romcal).

Likewise, the term "Martyrology" could indeed disappear. It's a term commonly used in monastic circles, but in practice, I found it to be confusing or not sufficiently explicit within the context of this Romcal project (the primary goal here being to add metadata to individuals, whether saints or blesseds).

**Reply by TobiTenno** [reply]
completely agree w/ fallbacks.

I'm not sure I understand how titles ties in, given that the localized data, right now, doesn't specifically mention that it's corresponding to "titles", that just happens to be how it's presently used, but i'd think the localization data should be usable in most contexts, and this is localized data about that Faithful Departed.

**Reply by emagnier** [reply]
Regarding the titles, I'm referring to metadata like this:
```yaml
titles: [Title.Priest, Title.Martyr],
```
It's already possible to define these titles, just like colors, within the calendar definitions. Currently, defining these titles in the martyrology file helps avoid duplications in each calendar. Unless the color is specifically indicated, Romcal can already deduce the liturgical color of a day, so these title metadata provide additional information to determine the correct color.

To come back to the broader topic, you're right, we need to consider the localization system and data structure to make them usable in the majority of contexts.

**Reply by tukusejssirs** [reply]
> [@TobiTenno]
>
> re: Names in martryology:
would anyone be opposed to either (1) moving these names to the locale files and expanding locales to be objects w/ `name`, `titledName`, and any other combinations we need to create over time; or (2) using the localized value from the locales files entirely and remove `martyrology#name` entirely?

I am for moving the names to locale files (see #154) and all other celebration metadata used to generate celebration names. This would replace `.names` in l10n files altogether. Therefore, from this point of view, I am for the first option you propose.

Moreover, I suggest localising first names separately, as they are used in multiple/most celebration names. However, we need to add an option to keep the original first name spelling and/or a way to override first name localisation, as in some locales some first names (esp of newer persons) are not localised at all.

Also note that we need to have multiple forms of the most localised data, as in some languages (other than English; e.g. in Slovak or Czech), we usually use [Genitive grammar case](https://en.wikipedia.org/wiki/Genitive_case) (`[Feast of] Some Celebration name`) for celebration names, however having the [Nominative grammar case](https://en.wikipedia.org/wiki/Nominative_case) is quite useful in other use cases. OFC this is locale-dependent and should be configurable in locale files (i.e. we default a default config) and the user should be able to override the config (just like we can for calendar configuration).

As for the second option you propose, I don’t think we should totally remove names, surnames and other data localised in l10n files from martyrology (or celebration metadata, as I call it now). They would be replaced by the l10n IDs, so that we could localise them during calendar generation.

> [@emagnier]
>
> We should establish a fallback system (similar to how it's already done with English) so that non-directly translatable metadata (like boolean data or dates, for example, such as the current `dateOfDeathIsApproximative: boolean` property) can automatically propagate into all bundles (packages containing calendar definitions and associated localization) generated during the library build.

I am not sure if I get it what you are talking about.

How should we back for boolean data or dates? IMHO dates are not translatable and we should require a default date format configuration in l10n files which.

As for boolean data, as I see it, you are talking about any metadata that has a boolean value (e.g. whether a celebration is a holy day of obligation). I believe that all such data (e.g. `holy day of obligation`) should be required to be localised, just like all data we have localised in `.seasons`, `.ranks`, `.colors` (and some others) in the l10n files. IMHO is should not be much of work to localise them in any locale, and when we use a localisation tool, we could check if all required strings are localised; if they are not, we would let the translators know somehow (e.g. by giving them a separate list of required strings in order to publish a locale) that a locale is missing some required strings, thus can cannot publish a new locale yet.

Unless I misunderstood your comment. :thinking:

> [@emagnier]
>
> However, it's important to note that the `titles` properties are currently used by Romcal to determine the liturgical color of the day (if the person is a martyr = red, otherwise white, or violet depending on the liturgical period). We need to figure out how to address this issue and find the right balance: either redefine titles or colors at the calendar definitions level (potentially duplicating the same metadata for the same entity); or preferably, look for this information in the localization file without requiring a localization library to fetch this specific information (I want to ensure that a localization library is not necessary for the core functioning of Romcal).

I am for keeping the automatic colour determination. :wink:

I didn’t want to suggest to neither remove titles nor the automatic colour determination. All I suggest is to move the titles to celebration metadata (martyrology if you wish). Then, if a celebration references an ID of a person with titles defined, all the titles defined in metadata for the person would be used, _unless_ the celebration specifies the titles; note that the celebration should be able to use only titles defined in metadata for the particular saint. And OFC, the localised title would get acquired from l10n files, similarly how it is done now.

> [@emagnier]
>
> Likewise, the term "Martyrology" could indeed disappear. It's a term commonly used in monastic circles, but in practice, I found it to be confusing or not sufficiently explicit within the context of this Romcal project (the primary goal here being to add metadata to individuals, whether saints or blesseds).

Yeah, while the term _martyrology_ is not only used in monastic circles (we have _Martyrologium Romanum_! :wink:), I too think we should replace it with something like _[celebration] metadata_. I agree that it might not be properly understood by all contributors, and even I think that the term does not completely describe the way we actually use it.

And in _celebration metadata_, I personally would put metadata of non-personal celebrations too, as well as collective names (celebration names of multiple persons with a distinct name, e.g. [`seven_holy_founders_of_the_servite_order`](https://github.com/romcal/romcal/blob/0b7aced379d62288db080744508bf5624c9e84dd/lib/locales/en.ts#L925)), although we might want to move them under a different property (e.g. `type CelebrationMetadata = {persons: Person[], nonPersonalCelebrations: NonPersonalCelebration[], collectiveNames: CollectiveName[]}`).

> [@TobiTenno]
>
> i'd think the localization data should be usable in most contexts, and this is localized data about that Faithful Departed.

I am not sure if I understand what you’re talking about. Why do you specifically talk about the Faithful Departed? :thinking:

**Reply by TobiTenno** [reply]
> [@tukusejssirs]
> I am not sure if I understand what you’re talking about. Why do you specifically talk about the Faithful Departed? 🤔

I think this was more in relation to another part that you already addressed, so don't worry about that. It was a general term for all those who are venerable, blessed, and canonized

**Reply by emagnier** [reply]
> How should we back for boolean data or dates? IMHO dates are not translatable and we should require a default date format configuration in l10n files which.

@tukusejssirs, that's exactly what I mentioned above: some data aren't directly translatable. However, they should still be included automatically in all localization files (within all localized calendars packages) as useful information for string composition or interpolation, or for anything one might want to create for a specific locale.

> I didn’t want to suggest to neither remove titles nor the automatic colour determination. 

@tukusejssirs, we agree on that. However, what I suggested is to define them in the English locale file rather than in the calendar definitions (and not in the martyrology, which won't exist anymore), to avoid scattering and redundant information.

And regarding the rest, @tukusejssirs, I agree with your replies.

**Reply by tukusejssirs** [reply]
> However, what I suggested is to define them in the English locale file rather than in the calendar definitions (and not in the martyrology, which won't exist anymore), to avoid scattering and redundant information.

@emagnier, I am not sure why martyrology would be removed altogether. IMHO it should be kept, but it could be renamed to _[celebration] metadata_ (or something similar). The purpose of the metadata is to keep all the metadata of saints, blesseds, non-personal celebrations in one place and reference them from celebration definitions in calendars. (I was under the inpression we understood each other on this. :thinking:)

IMHO, the data should be split to the following independent types of data which would link data items of other types of data by ID:

- sources;
- celebration metadata (was: martyrology);
- calendars with celebrations;
- localisation.

No localisation data should be in any other type of data.

**Comment by emagnier** [comment]
**Localized calendar packages:**

Thank you @tukusejssirs for bringing up this discussion. I'd like to delve deeper into the topic of the format and distribution of our final files, particularly focusing on what we currently call "bundle files". These compiled files are generated from a build step, to contain only the necessary data for a calendar (calendar definitions + parent calendars combined), in a specific locale (including only the necessary locale entries and martyrology metadata). Everything is exported as JS files in various formats (CJS, ESM, IIFE) and distributed as NPM packages.

Originally, this idea stemmed from the concept of being able to use Romcal without necessarily integrating all possible localizations, aiming to keep the Romcal library's size reasonable when imported and integrated into other projects. Another consideration is performance, pre-computing certain calendar definition data to avoid recalculating it with every request to Romcal. And the code managing this build initially started as a poc, somewhat loosely written and tested.
@TobiTenno, you likely have ideas or experience to share here. I believe the build step remains necessary, but is it really necessary to package these as NPM packages? Especially with the proliferation of calendars and locales, this will only continue to grow in the future. And is it worthwhile to push file size optimizations to the extreme, considering that Romcal will likely be integrated into projects that already have their own build and optimization systems?

I've already conducted some experiments in my `optimizations-ex` branch, where I revisited the model and distributed data format, and where the bundles are no longer packaged as NPM packages but as plugins distributed within the main Romcal NPM package. This approach seems ultimately simpler to use and to maintain.

In any case, there are 3 points I'd like to see evolve: (1) distributing the built data as JSON (instead of JS), (2) associating a JSON Schema with this distributed data, and (3) having tools and a packaging architecture that remain as simple as possible to use, maintain, and test.

**Reply by TobiTenno** [reply]
I think a decent tree-shaker would eliminate extra plugins in bundling. Keeping in mind that we have examples wherein we provide an api, which someone might want to make very specific to a single calendar; I think we should still distribute them separately, and given that we already have the whole scope, i think it makes sense to separate them how we do now. 

That said, I might agree that using a JSON schema so that the types are a bit simpler would be nice, but that would also complicate consuming the types, compared to now, given that json schema-based types aren't as intuitive as readily typed packages.

The 3rd evolution aspect, I think, is very sustainable apart from changing what these distributions are; i.e. js vs json, module vs plugin; and we could very practically add more tests, some being things like testing translations existing, or that everything in translations exists in `en.ts`. Those are pretty big practical steps we can take, and iteratively add tests. Adding iteratively, keeping what we want broken down to a small degree so that we don't try to implement Rome all in one day and avoid burnout, allows more external contributors to join in on it as well, and gives pretty clear things to fix, like defining a test-coverage aspect we want to add.

My belief on that part would be broken out as an issue with:
- Thing we want to test -> How we want to test it

**Reply by tukusejssirs** [reply]
> [@emagnier]
>
> I'd like to delve deeper into the topic of the format and distribution of our final files, particularly focusing on what we currently call "bundle files". These compiled files are generated from a build step, to contain only the necessary data for a calendar (calendar definitions + parent calendars combined), in a specific locale (including only the necessary locale entries and martyrology metadata). Everything is exported as JS files in various formats (CJS, ESM, IIFE) and distributed as NPM packages.

To be honest, ATM I am concentrating on the final/target data types/structure. That said, the bundling is equally important.

I think that the sources, celebration metadata and l10n files (all of these I call collectively as the _[romcal] data_) should be as DRY as possible, however, rather be a bit repetitive than remove any part of the data. However, during the build step, IMO we should output only the required parts of the data (just like we do it right now).

> [@emagnier]
>
> Originally, this idea stemmed from the concept of being able to use Romcal without necessarily integrating all possible localizations, aiming to keep the Romcal library's size reasonable when imported and integrated into other projects.

I agree with this, however, for sites like romcal.js.org, it is potentially not a good idea to reduce the data, as most of the data can be used after selecting different calendar or locale. (I’m just thinking aloud.)

> [@emagnier]
>
> Another consideration is performance, pre-computing certain calendar definition data to avoid recalculating it with every request to Romcal.

Again, I think that pre-computing all the necessary data in the build step is the way to go. However, the data (apart from the l10n data) should still be locale-agnostic. Also, the calendars/celebrations should not define any new celebration metadata unless necessary (i.e. it is variable, one way or another; e.g. holy days of obligation need to be defined in celebrations, as they vary from calendar to calendar).

> [@emagnier]
>
> is it really necessary to package these as NPM packages?

While it might not be necessary to publishe all metadata, calendars, and locales as separate NPM packages, I thought the idea was to let the user import what they want to use without importing large libraries.

Also, when we define the data in a code-agnostic way (i.e. in JSON and YAML), we wanted to let other libraries/apps (outside romcal) to use the data, thus it _might_ be useful to publish these packages. On the other hand, if the library/app is written in a different programming language, NPM packages don’t matter at all.

> [@emagnier]
>
> I've already conducted some experiments in my `optimizations-ex` branch, where I revisited the model and distributed data format, and where the bundles are no longer packaged as NPM packages but as plugins distributed within the main Romcal NPM package. This approach seems ultimately simpler to use and to maintain.

Plugins could be a good idea; however, will romcal then automatically download the required plugins (with the data) at runtime? Or how will we distribute the plugins? :thinking:

> [@emagnier]
>
> In any case, there are 3 points I'd like to see evolve: (1) distributing the built data as JSON (instead of JS), (2) associating a JSON Schema with this distributed data, and (3) having tools and a packaging architecture that remain as simple as possible to use, maintain, and test.

:+1:

> [@TobiTenno]
>
> That said, I might agree that using a JSON schema so that the types are a bit simpler would be nice, but that would also complicate consuming the types, compared to now, given that json schema-based types aren't as intuitive as readily typed packages.

IMHO, the data (sources, metadata, calendars with celebrations, locales) could be stored as TypeScript code in the repository and then built into JSON/YAML. If we do it this way, we could move the types into an internal library where the types would be defined and thus consumed not only by the data libraries, but also by `@romcal/core`. Still, the data should be consumed in JSON/YAML by `@romcal/core`.

As I see it, we would use the JSON schema to validate the data, however, then we parse the data to objects of TS types imported from the internal library.

> [@TobiTenno]
>
> The 3rd evolution aspect, I think, is very sustainable apart from changing what these distributions are; i.e. js vs json, module vs plugin

I’ll leave the discussion on JS vs JSON vs YAML and module vs plugin for you two to find the way.

> [@TobiTenno]
>
> we could very practically add more tests, some being things like testing translations existing, or that everything in translations exists in `en.ts`. Those are pretty big practical steps we can take, and iteratively add tests. Adding iteratively, keeping what we want broken down to a small degree so that we don't try to implement Rome all in one day and avoid burnout, allows more external contributors to join in on it as well, and gives pretty clear things to fix, like defining a test-coverage aspect we want to add.
>
> My belief on that part would be broken out as an issue with:
> - Thing we want to test -> How we want to test it

Basically, I agree with you on all points here.

Re testing: I think we should try to find a way to thoroughly test all the data, all calendars, and all locales. For e2e tests, we should try to test all locales for an extended period of time (i.e. several years) in which we would test general stuff and specific issues. ATM, we do this only sporadically, but definitely not thoroughly enough.

Even some statistics generation (e.g. what percentage of `en.ts` strings are localised in each locale) or calendar propers and putting it into the `docs` app after each commit would be nice.

**Reply by TobiTenno** [reply]
on all the replies to me, I agree with you from what I'm reading 👍

**Reply by emagnier** [reply]
Thank you both for your feedback. I also believe that the current packaging approach for localized calendar packages is the right one (good initial intuition, after all). ++ for using a tree-shaker to eliminate extra plugins in bundling!

Regarding distribution, when I mentioned NPM packages or distributing as a "plugin" within the main NPM package (similar to what dayjs does, for example), it's just a matter of distribution method, as the usage remains exactly the same. I'm still undecided on this point for now.

Lastly, I agree with you @TobiTenno on distributing these packages as JS files (as is currently done). The idea of distributing them as JSON + JSON Schema might be a bit idealistic for now, as using JS files directly is actually much simpler, and with native typing. If a real concrete use case requiring JSON arises in the future, then we can evolve accordingly, but let's keep it simple for now.

**Reply by tukusejssirs** [reply]
> Regarding distribution, when I mentioned NPM packages or distributing as a "plugin" within the main NPM package (similar to what dayjs does, for example), it's just a matter of distribution method, as the usage remains exactly the same. I'm still undecided on this point for now.

Thanks for the explanation, Etienne! :pray:

> Lastly, I agree with you @TobiTenno on distributing these packages as JS files (as is currently done). The idea of distributing them as JSON + JSON Schema might be a bit idealistic for now, as using JS files directly is actually much simpler, and with native typing. If a real concrete use case requiring JSON arises in the future, then we can evolve accordingly, but let's keep it simple for now.

I am okay with that, however, it might be still a good idea to provide JSON/YAML files (as a build artefact), even though romcal won’t use it. :thinking:

**Reply by emagnier** [reply]
I'm trying to summarize our discussions on this topic: 

1. **Build & Export Format**:
   - Agreement to distribute packages as JS files for simplicity and native typing (CJS, ESM, IIFE formats).
   - Potentially consider adding a 4th format in JSON (+ JSON Schema).

2. **Distribution**:
   - Currently distributed as separate NPM packages (keeping this method for now).

3. **Technical Considerations**:
    - Exploring the use of a tree-shaker tool to eliminate unnecessary content in localized calendar packages could significantly simplify the build process, as opposed to constructing them as currently done.
   - Add some tests to ensure the build and packaging processes work correctly.
   - Data validation itself should also be performed, but separately from the build + packaging process (separation of concerns).

**Reply by TobiTenno** [reply]
ah, the tree-shaking was meaning consumers should leverage something similar to avoid packaging more than they want, but not all of them work great with bundled files.

what kind of data validation do you mean? just checking, because if you just mean the kind of things we'd do in a CI before starting build + package, then I agree, but if you mean something like that would only be run as a manual check locally for a developer, i think we'd need to integrate that into CI still

**Reply by tukusejssirs** [reply]
I have updated the OP and added the @emagnier’s [comment](https://github.com/romcal/romcal/discussions/528#discussioncomment-9218420) there, as I think we need to summarise the outcome of our discussions in the OP and finally in the `docs` app (in contribution or development section).

**Comment by tukusejssirs** [comment]
I have just tried to run `pnpm data:checks`, `pnpm docs:check-links` and `pnpm docs:sort-glossary` and all fail to run due to missing dependencies, then I installed the deps used in `pnpm docs:check-links` and it all worked as expected. Finally, I noticed there are five `package.json` files:

- `package.json`;
- `packages/config/package.json`;
- `packages/easter/package.json`;
- `rites/roman1962/package.json`;
- `rites/roman1969/package.json`.

In `rites/roman1969/package.json`, the deps are defined, but why? Either way, I think we should be able to install all the deps using a single command (e.g. `pnpm i`) and run an NPM script (e.g. `pnpm docs:check-links`) from anywhere in the repository.

I think that the docs should be separate package (an app; it does not need to be released on NPM) which should contain all the docs of all the packages (apps, libs and [internal] tools) which are part of the monorepo. I also think that all packages should be version together, so that we can always tell that a specific version (e.g. `v3.0.1`) of all packages will work together (yes, that also means that we would release version of packages which have no changes when others will have some changes).

Finally, I still think that Nx monorepo is much better. I also think we should use an integrated monorepo when there is a single `package.json` file where we store all the dependencies. This way, we can always use the same dependency versions across all packages (apps/libs/tools).

I find the current state very confusing and I also see many shortcomings which are already solved in monorepo ‘frameworks’ like Nx.

**Reply by tukusejssirs** [reply]
Another reason why Nx is a better chose: it is more package manager agnostic. Right now, I get the following warning when using PNPM:

```bash
 WARN  The "workspaces" field in package.json is not supported by pnpm. Create a "pnpm-workspace.yaml" file instead.
```

**Reply by TobiTenno** [reply]
why should it be package manager agnostic? I've yet to hear why package manager choice is necessary. we build npm package locks, we publish on npm.

this sounds like a case of pnpm not properly resolving packages the way node does. if node's resolution is the defacto-default, I disagree that we should try to make our system use a special monorepo that any of the numerous custom package managers can use, but that the numerous package managers should try to not break npm's standards.
